### PR TITLE
SPU2: Add Cubeb backend, remove Portaudio and SDL2 backends

### PR DESCRIPTION
**Pull request by stenzek**
### Description of Changes
Another change from the Qt branch.

Everything should be functional except device selection, which still needs to be added. But do we care? Linux didn't have it before.

Cubeb also handles device hotplugging quite nicely :)

The last commit adds a pause method for the device. Currently, this isn't used, but in Qt we call it when the VM is paused, since we don't tear down everything there, we don't want to be burning CPU cycles/battery when we're paused.

Currently the Windows default is still XAudio. Did we want to change this?

### Rationale behind Changes

Consistency across platforms. Windows used portaudio/xaudio, Linux used portaudio or sdl, and Mac used sdl. This puts all platforms on a level playing field, although I left the xaudio backend for Windows in place because it's not really hurting anything.

### Suggested Testing Steps

 - Make sure audio works (I've done this on Windows/Linux).
 - Make sure surround expansion works (I can't test this myself).

### CODE REVIEW

**Comment by kenshen112** [comment]
So, one concern I've had with cubeb cause I've had a working branch and so has Ziemas at one point or another is that it seems to require a dependency on rust with backends like `pulse-rust`

**Comment by stenzek** [comment]
The rust backends are disabled... (https://github.com/PCSX2/pcsx2/pull/5183/files#diff-9dae0716d278e398f7b4675b70f8560505b1bc917f4c180591f46c9f554f08b7)

File: pcsx2/SPU2/SndOut_Cubeb.cpp [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+			default:
+				channels = 2;
+				break;
+		}
+
+		switch (channels)
+		{
+			case 2:
+				ConLog("* SPU2 > Using normal 2 speaker stereo output.\n");
+				ActualReader = new ConvertedSampleReader<StereoOut16>(&writtenSoFar);
```
Comment (Warepire) [inline pcsx2/SPU2/SndOut_Cubeb.cpp]: I don't see a `delete` anywhere, any risk this could end up leaking?

Comment (stenzek) [reply pcsx2/SPU2/SndOut_Cubeb.cpp]: True, this can be a unique_ptr.

**Comment by kenshen112** [comment]
> The rust backends are disabled... (https://github.com/PCSX2/pcsx2/pull/5183/files#diff-9dae0716d278e398f7b4675b70f8560505b1bc917f4c180591f46c9f554f08b7)

Looks good. Will be super nice to finally have cubeb in this project 👍

File: pcsx2/SPU2/Windows/SPU2.rc [hunk 2]
```diff
@@ -175,25 +174,22 @@ LANGUAGE LANG_ENGLISH, SUBLANG_ENGLISH_UK
 // Dialog
 //
 
-IDD_PORTAUDIO DIALOGEX 0, 0, 316, 166
+IDD_CUBEB DIALOGEX 0, 0, 316, 166
 STYLE DS_SETFONT | DS_MODALFRAME | DS_FIXEDSYS | WS_POPUP | WS_CAPTION | WS_SYSMENU
 CAPTION "Portaudio Output Module Settings"
```
Comment (MrCK1) [inline pcsx2/SPU2/Windows/SPU2.rc]: Missed the window title here

File: 3rdparty/cubeb/include/cubeb_export.h [hunk 3]
```diff
@@ -0,0 +1,25 @@
+/*  PCSX2 - PS2 Emulator for PCs
+ *  Copyright (C) 2002-2020  PCSX2 Dev Team
```
Comment (lightningterror) [inline 3rdparty/cubeb/include/cubeb_export.h:2]: 2021

**Review by lightningterror — COMMENTED** [review]
I see some codacy complaints which can be resolved.

File: pcsx2/SPU2/SndOut_Cubeb.cpp [hunk 4]
```diff
@@ -0,0 +1,341 @@
...
+
+		// TODO: Once the config stuff gets merged, drop the wxString here.
+		wxString backend;
+		CfgReadStr(L"Cubeb", L"BackendName", backend, L"");
+		m_Backend = StringUtil::wxStringToUTF8String(backend);
+
+		if (m_SuggestedLatencyMS < 10)
+			m_SuggestedLatencyMS = 10;
+		if (m_SuggestedLatencyMS > 200)
+			m_SuggestedLatencyMS = 200;
```
Comment (refractionpcsx2) [inline pcsx2/SPU2/SndOut_Cubeb.cpp]: Not huge, but could be replaced with std::clamp

**Comment by stenzek** [comment]
> I see some codacy complaints which can be resolved.

Last two here are false positives (it's obviously not considering the macro).

**Comment by gilderoylockhart01** [comment]
If I am being honest, is it worth keeping xaudio still since we now have a fully working cubeb backend?

**Comment by refractionpcsx2** [comment]
We can tear it out later if it's no longer required.

**Review by refractionpcsx2 — APPROVED** [review]

**Review by Mrlinkwii — APPROVED** [review]

**Comment by stenzek** [comment]
Cubeb doesn't have an XAudio backend, and XAudio is useful later on if we did a UWP port (it's the only option there).

But I think changing the default is still a good idea.

**Comment by gilderoylockhart01** [comment]
Does UWP require a specific Xaudio version like directX? (it needs 12 if I remember)

**Comment by stenzek** [comment]
DX11 works too but you can only get feature level 11/12 via dx12. It's just xaudio2, and the sdk takes care of all the versioning.

**Comment by gilderoylockhart01** [comment]
Wow, that's pretty damn impressive, can't wait we get rid of that WX nightmare so that makes that possible. Imagine the thing, PCSX2 running on the Xbox Series X/S.

**Review by RedDevilus — APPROVED** [review]
Atleast it doesn't need a hack for handling 24 bit 96000 Hz for my cheap headset that PortAudio needed. Seems to work fine.  :)

**Comment by RedDevilus** [comment]
frames / 48

480/48 = 10ms (my PC)
1200/48 = 25ms (link PC) which is above buffer size but has lower specced PC anyway

Might be best to have Cubeb as default like you said. Kinda funny comparing 1.6 Linux 300 Ms to now such low amounts of ms.

**Comment by stenzek** [comment]
Keep in mind that's just the output latency. There's still another level of buffering in there, and considering you're sleeping up to 16ms per frame ;) but yeah, should be pretty decent overall I hope, and certainly no worse than the current setup.

**Comment by 987123879113** [comment]
I primarily play music/rhythm games so I wanted to test how this affected latency as I usually rely on the PortAudio backend, but I'm getting a crash when using the WASAPI backend on Windows 11 now. Multiple runs show a different crash address but the common point of failure seems to be inside Cubeb. WinMM doesn't seem affected. I tried playing around with the latency sliders to see if it changed anything but it didn't seem to help.

After testing various audio devices and changing the device settings, it crashes on every device when it's not using 48 kHz sample rate. Bit depth doesn't seem to be a factor.

Onboard (2ch, 16bit, 44.1kHz): Crash
Onboard (2ch, 16bit, 48kHz): No crash
Onboard (2ch, 24bit, 44.1kHz): Crash
Onboard (2ch, 24bit, 48kHz): No crash

Apple USB-C to 3.5mm Headphone Jack Adapter (2ch, 16bit, 44.1kHz): Crash
Apple USB-C to 3.5mm Headphone Jack Adapter (2ch, 16bit, 48kHz): Working
Apple USB-C to 3.5mm Headphone Jack Adapter (2ch, 24bit, 44.1kHz): Crash
Apple USB-C to 3.5mm Headphone Jack Adapter (2ch, 24bit, 48kHz): Working

Focusrite Scarlett 2i4 1st Gen (16bit 44.1kHz): Crash
Focusrite Scarlett 2i4 1st Gen (16bit 48kHz): Working
Focusrite Scarlett 2i4 1st Gen (24bit 44.1kHz): Crash
Focusrite Scarlett 2i4 1st Gen (24bit 48kHz): Working

Yamaha AG-06 (2ch, 24bit, 44.1kHz): Crash
Yamaha AG-06 (2ch, 24bit, 48kHz): Working
Yamaha AG-06 (2ch, 24bit, 88.2kHz): Crash
Yamaha AG-06 (2ch, 24bit, 96kHz): Crash
Yamaha AG-06 (2ch, 24bit, 176.4kHz): Crash
Yamaha AG-06 (2ch, 24bit, 192kHz): Crash

The actual error itself changes often but it's always a heap corruption around the same `cubeb_resampler_speex_one_way<short>::input_buffer(unsigned __int64 frame_count)` part of the code.

```
Debug Error!

Program: D:\Projects\pcsx2\bin\pcsx2x64-avx2-dbg.exe

HEAP CORRUPTION DETECTED: after Normal block (#935496) at 0x0000027305ED9850.
CRT detected that the application wrote to memory after end of heap buffer.


(Press Retry to debug the application)
Critical error detected c0000374
pcsx2x64-avx2-dbg.exe has triggered a breakpoint.

```

Call stack (1):
```
>	pcsx2x64-avx2-dbg.exe!operator new(unsigned __int64 size) Line 35	C++
 	pcsx2x64-avx2-dbg.exe!operator new[](unsigned __int64 size) Line 29	C++
 	pcsx2x64-avx2-dbg.exe!auto_array<short>::reserve(unsigned __int64 new_capacity) Line 164	C++
 	pcsx2x64-avx2-dbg.exe!cubeb_resampler_speex_one_way<short>::input_buffer(unsigned __int64 frame_count) Line 311	C++
 	pcsx2x64-avx2-dbg.exe!cubeb_resampler_speex<short,delay_line<short>,cubeb_resampler_speex_one_way<short>>::fill_internal_output(short * input_buffer, long * input_frames_count, short * output_buffer, long output_frames_needed) Line 184	C++
 	pcsx2x64-avx2-dbg.exe!cubeb_resampler_speex<short,delay_line<short>,cubeb_resampler_speex_one_way<short>>::fill(void * input_buffer, long * input_frames_count, void * output_buffer, long output_frames_needed) Line 164	C++
 	pcsx2x64-avx2-dbg.exe!cubeb_resampler_fill(cubeb_resampler * resampler, void * input_buffer, long * input_frames_count, void * output_buffer, long output_frames_needed) Line 360	C++
 	pcsx2x64-avx2-dbg.exe!`anonymous namespace'::refill(cubeb_stream * stm, void * input_buffer, long input_frames_count, void * output_buffer, long output_frames_needed) Line 829	C++
 	pcsx2x64-avx2-dbg.exe!`anonymous namespace'::refill_callback_output(cubeb_stream * stm) Line 1215	C++
 	pcsx2x64-avx2-dbg.exe!`anonymous namespace'::wasapi_stream_render_loop(void * stream) Line 1355	C++
 	[External Code]	


```

Call stack (2):
```
>	pcsx2x64-avx2-dbg.exe!operator delete(void * block) Line 38	C++
 	pcsx2x64-avx2-dbg.exe!operator delete[](void * block) Line 32	C++
 	pcsx2x64-avx2-dbg.exe!auto_array<short>::reserve(unsigned __int64 new_capacity) Line 170	C++
 	pcsx2x64-avx2-dbg.exe!cubeb_resampler_speex_one_way<short>::input_buffer(unsigned __int64 frame_count) Line 311	C++
 	pcsx2x64-avx2-dbg.exe!cubeb_resampler_speex<short,delay_line<short>,cubeb_resampler_speex_one_way<short>>::fill_internal_output(short * input_buffer, long * input_frames_count, short * output_buffer, long output_frames_needed) Line 184	C++
 	pcsx2x64-avx2-dbg.exe!cubeb_resampler_speex<short,delay_line<short>,cubeb_resampler_speex_one_way<short>>::fill(void * input_buffer, long * input_frames_count, void * output_buffer, long output_frames_needed) Line 164	C++
 	pcsx2x64-avx2-dbg.exe!cubeb_resampler_fill(cubeb_resampler * resampler, void * input_buffer, long * input_frames_count, void * output_buffer, long output_frames_needed) Line 360	C++
 	pcsx2x64-avx2-dbg.exe!`anonymous namespace'::refill(cubeb_stream * stm, void * input_buffer, long input_frames_count, void * output_buffer, long output_frames_needed) Line 829	C++
 	pcsx2x64-avx2-dbg.exe!`anonymous namespace'::refill_callback_output(cubeb_stream * stm) Line 1215	C++
 	pcsx2x64-avx2-dbg.exe!`anonymous namespace'::wasapi_stream_render_loop(void * stream) Line 1355	C++
 	[External Code]	

```

**Comment by refractionpcsx2** [comment]
44100 won't work because that's below the required samplerate for the emulation, so working or not in those scenarios is irrelevant really.

As for the latency, it reports it's working fine at 10ms (plus 15ms internal) and the audio seems good.

**Comment by 987123879113** [comment]
Gotcha. My audio interfaces were defaulting to 44.1kHz so I wasn't aware it was a requirement until I tested this PR. I recently upgraded and started using an audio interface again so I never noticed an issue with 44.1kHz prior to this and thought it might be the change itself. Sorry stenzek.

I guess an issue now becomes the hard crash when using an unsupported sample rate. I just tested with current nightly PCSX2 and it shows an error in the program log and then continues playing the game without audio.
```
Failed to create Port Audio Device 48khz, Please use Exclusive Mode
* SPU2: PortAudio error: Invalid sample rate
```

With the Cubeb backend, the debug build shows the heap corruption message before crashing and the release build silently hard closes PCSX2 without any feedback.

**Comment by tellowkrinkle** [comment]
> 44100 won't work because that's below the required samplerate for the emulation

Resampling is a thing, there's no reason we couldn't support it.

But either way, we should not be crashing if the user's audio output is using a lower sample rate than the emulator

**Comment by refractionpcsx2** [comment]
yeah the main takeaway is the crashing shouldn't happen, but cubeb should be able to resample at least downward.

**Comment by stenzek** [comment]
Most of my testing was with an older version of cubeb, not the current upstream master, so it's very possible there's an issue here. I'll see if I can reproduce it by forcing the rate..

File: pcsx2/SPU2/SndOut_Cubeb.cpp [hunk 5]
```diff
@@ -0,0 +1,341 @@
...
+				channels = 8; // we do not support 7.2 or more, downgrade to 7.1
+				break;
+		}
+
+		cubeb_stream_params params = {};
+		params.format = CUBEB_SAMPLE_S16LE;
+		params.rate = SampleRate;
+		params.channels = channels;
+		params.layout = CUBEB_LAYOUT_UNDEFINED;
+		params.prefs = CUBEB_STREAM_PREF_PERSIST;
```
Comment (arcum42) [inline pcsx2/SPU2/SndOut_Cubeb.cpp]: In cubeb.h, it mentions that CUBEB_STREAM_PREF_PERSIST is obsolete and ignored by all backends. We could probably just use CUBEB_STREAM_PREF_NONE instead...

**Comment by stenzek** [comment]
Okay, it looked like there was some possible buffer overruns when the SndBuffer underflowed. Those should be fixed now. Funnily enough it was kinda jank before, it was overwriting partial packets with silence.. guess nobody noticed :)

I've tested at 44100 with appverifier on, and no overruns.

Also changed the message to log the latency in milliseconds as well as frames, and set the correct speaker layout for >2 speaker setups. But I can't test that.

**Comment by RedDevilus** [comment]
Sorry, I can't hear silence as well as tones and general sounds.

File: pcsx2/pcsx2.vcxproj.filters [hunk 6]
```diff
@@ -2915,4 +2913,4 @@
       <Filter>AppHost\Resources</Filter>
     </Manifest>
   </ItemGroup>
-</Project>
```
Comment (tellowkrinkle) [inline pcsx2/pcsx2.vcxproj.filters]: Leave the newline

File: pcsx2/pcsx2.vcxproj.filters [hunk 7]
```diff
@@ -2753,6 +2750,7 @@
     <ClInclude Include="PerformanceMetrics.h">
       <Filter>System</Filter>
     </ClInclude>
+    <ClInclude Include="GS.h" />
```
Comment (tellowkrinkle) [inline pcsx2/pcsx2.vcxproj.filters]: Keep the filter on

Comment (stenzek) [reply pcsx2/pcsx2.vcxproj.filters]: Ugh how did that make it in >.>

**Review by tellowkrinkle — APPROVED** [review]

**Comment by refractionpcsx2** [comment]
@987123879113 can you retest please.

**Comment by 987123879113** [comment]
@refractionpcsx2 @stenzek Tested against the latest commit `6c35c09fd93e9792172c176f3eb6ba5bf88ee7f6`. Looks good to me. No more crashing, and it's handling resampling properly for all sample rates I tested using the WASAPI backend with my Yamaha AG-06 at 44.1kHz, 48kHz, 88.2kHz, 96kHz, 176.4kHz, and 192kHz.

Just to be sure that there wasn't some trickery with my system going on I threw in a call to `cubeb_get_preferred_sample_rate` to display what sample rate Cubeb was seeing my device as being configured for and it properly reported the sample rate I expected for each test. 👍

**Comment by stenzek** [comment]
Thanks for checking such a wide range of setups!

**Comment by trostboot** [comment]
@stenzek Just for reference, audio expansion works as before with both 4ch and 5.1 layouts (can't test 7.1 myself) with either backend.

**Pull request merged**